test_polygon.py fails sporadically with ContractNotFound

### How to reproduce locally

```console
# 1. ensure that the keys below hold MATIC in Polygon network

# 2. set your own keys
export REMOTE_TEST_PRIVATE_KEY1=0x8c1..
export REMOTE_TEST_PRIVATE_KEY2=0xc217..

# 3. install pytest-repeat
pip install pytest-repeat

# 4. run the test. Repeat the test up to 100 times (count=100), until failure (-x). Typical: 25 runs until fail
pytest --count=100 -x tests/integration/remote/test_polygon.py::test_ocean_tx__create_data_nft
```

### Failures observed

1. When calling `DataNFT(self.config_dict, address)`, error is `ContractNotFound(f"No contract deployed at {address}")`. [CI log](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795). [Local log](https://github.com/oceanprotocol/ocean.py/files/10112502/local_log.txt).

2. When calling `data_nft.symbol()`, error is `ValueError: No data was returned - the call likely reverted`. [CI log](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656)

3. When calling `nft_factory.deployERC721Contract()`, error is `ValueError: nonce too low`. [Local log](https://github.com/oceanprotocol/ocean.py/files/10113084/nonce.txt).

I've observed (1) many times, and can repeat locally. I've only observed (2) once, and can't repeat. So let's focus on (1). I observed (3) a few times in a row, locally, when running repeatedly in quick succession.

### Simple trace of error 1

```text
-> tests/integration/remote/util.py:96: in do_ocean_tx_and_handle_gotchas
-> data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
-> ocean_lib/ocean/ocean.py:157: in create_data_nft
-> token = DataNFT(self.config_dict, address)

Returns error:
ContractNotFound(f"No contract deployed at {address}")
```
- The address is 0x32a1e8152d0171444f3c982E83E6467fe8c0aa34
- Polygon _does_ have it, as a contract, as expected. [Here](https://polygonscan.com/address/0x32a1e8152d0171444f3c982E83E6467fe8c0aa34#code)

The appendix has full tracebacks for error 1 and error 2.

### Chronology of errors observed

- In [PR#1124](https://github.com/oceanprotocol/ocean.py/pull/1124): Asset class -> DDO class, and asset.py -> ddo.py
  - CI tests passed
- Then barge 3.0 update came out (ref Alex in slack). "Expect instability" 
- I merged PR#1124 to main, in commit https://github.com/oceanprotocol/ocean.py/commit/091298a293f19ee58b32eee3e65da63861b62a0e
  - [CI tests failed](https://github.com/oceanprotocol/ocean.py/actions/runs/3569135420/jobs/5998769326). Error: "Error: No such container: ocean_aquarius_1". My guess: this was due to barge instability. But that may have masked the test_polygon ContractNotFound error
- Then I did a new branch from main, with [PR#1126](https://github.com/oceanprotocol/ocean.py/pull/1126). Small change: in download_asset_files(), change the first argument name from asset --> ddo.
  - CI tests passed
- Then I did a "touch" commit on main (https://github.com/oceanprotocol/ocean.py/commit/6e4620dde6a2bd16c0642a61f9510b8dabceb67d), in order to trigger CI
  - [CI tests failed](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250). Error: test_polygon ContractNotFound
- I merged PR#1126 to main
  - [CI tests failed](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251). It has 50+ lines of `Waiting for confirmation...`. The error is in test_polygon, but with `ValueError("No data was returned - the call likely reverted")`
- I ran the test locally 5 times. It failed once, with `ContractNotFound`. This is good news: I can reproduce locally!
- I installed pytest-repeat, and re-ran the test to keep going until a failure. In the first run, it took 25 tries to get a failure.

### Hypotheses

Most likely first
- It can take several blocks before the network confirms the tx. If confirmation doesn't happen (ie a "confirmation timeout"), then the Python code proceeds anyway. And that's where the error is
  - The call to create the data NFT _did_ go through, but it wasn't confirmed by the time that we tried to retrieve the DataNFT object from the address. But why?
- ..
- The call to create the data NFT did _not_ go through. But somehow the test kept proceeding anyway. Maybe there wasn't enough gas, and the try/except was failing
- Polygon network is unstable sometimes. (But: this is vague..)
- Something in a recent change on try/except related to enough MATIC funds in remote tests broke this (or another bug fix exposed this? We did have a bug fix yesterday shortly before this was observed)  
- Something in a recent change on asset -> ddo broke this

Q: Why does it fail more often on CI than locally?
- Hypothesis: my local brownie setting for # blocks to confirm is higher than CI's brownie setting
- Datapoint: in my local `~/.brownie/brownie-config.yaml`, "polygon" section, there is: `required_confs: 15`
- Datapoint: in ocean.py repo, [network-config.sample](https://github.com/oceanprotocol/ocean.py/blob/main/network-config.sample), there is: `required_confs: 15`. And, CI uses it: `.github/workflows/pytest.yml` copies it to `~/`, which automatically gets picked up by brownie
- Datapoint summary: both local & CI use 15 confirmations. Therefore the hypothesis is wrong.

Q: Does # confirmations affect how often the failure happens
- Hypothesis: yes
- Experiment: locally, set `required_confs` to 1, then observe rate of failure
  - Observed: For each run, number of repeats before fail: 20, 13, 5, 1, 4. (Excluding "Nonce too low" errors). Average: 8.6
- Experiment: locally, set `required_confs` to 15, then observe rate of failure
  - Observed: For each run, number of repeats before fail: 25, 1, 6, 5, 3. Average: 8
- Discussion: having more confirmations does not lead to a statistically significant difference in failure rate. Therefore the hypothesis is wrong.  

### Appendix: Full trace: error 1

From [this log](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795).

```text
________________________ test_ocean_tx__create_data_nft ________________________

tmp_path = PosixPath('/tmp/pytest-of-runner/pytest-0/test_ocean_tx__create_data_nft1')

    @pytest.mark.integration
    def test_ocean_tx__create_data_nft(tmp_path):
        """On Polygon, do a simple Ocean tx: create_data_nft"""
        # setup
        connect_to_network("polygon")
        util.set_aggressive_gas_fees()
    
        config = util.remote_config_polygon(tmp_path)
        ocean = Ocean(config)
    
        accounts.clear()
        (alice_wallet, _) = util.get_wallets()
    
        # Do a simple-as-possible test that uses ocean stack, while accounting for gotchas
>       util.do_ocean_tx_and_handle_gotchas(ocean, alice_wallet)

tests/integration/remote/test_polygon.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/integration/remote/util.py:96: in do_ocean_tx_and_handle_gotchas
    data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
ocean_lib/ocean/ocean.py:157: in create_data_nft
    token = DataNFT(self.config_dict, address)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
ocean_lib/web3_internal/contract_base.py:38: in __init__
    self.contract = load_contract(self.contract_name, address)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
ocean_lib/web3_internal/contract_utils.py:39: in load_contract
    return Contract.from_abi(contract_name, address, abi)
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/brownie/network/contract.py:969: in from_abi
    _DeployedContractBase.__init__(self, address, owner, None)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("Contract 'ERC721Template' object has no attribute 'address'") raised in repr()] Contract object at 0x7fd5812a4dc0>
address = '0x32a1e8152d0171444f3c982E83E64[67](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795#step:7:68)fe8c0aa34', owner = None, tx = None

    def __init__(
        self, address: str, owner: Optional[AccountsType] = None, tx: TransactionReceiptType = None
    ) -> None:
        address = _resolve_address(address)
        self.bytecode = web3.eth.get_code(address).hex()[2:]
        if not self.bytecode:
>           raise ContractNotFound(f"No contract deployed at {address}")
E           brownie.exceptions.ContractNotFound: No contract deployed at 0x32a1e8152d0171444f3c982E83E6467fe8c0aa34

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/brownie/network/contract.py:708: ContractNotFound
----------------------------- Captured stdout call -----------------------------
alice_wallet.address = '0x8bE6c8126C4383745c25F3b3Ca76D73E3Caf8544'
bob_wallet.address = '0xD6764a9F3387B14692CC783BcF5604a23EBb779f'
Call create_data_nft(), and wait for it to complete...

Transaction sent: 0x74b6bfab7ac0193b6b9af03fe934a085d2be8d40cd6d0ee[68](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795#step:7:69)923437d77e4c14d
  Max fee: 35.03839[69](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795#step:7:70)08 gwei   Priority fee: 35.0360738 gwei   Gas limit: 665075   Nonce: 164
  Waiting for confirmation... -
  Waiting for confirmation... \
  Waiting for confirmation... |
  Waiting for confirmation... /

  ERC[72](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795#step:7:73)1Factory.deployERC721Contract confirmed   Block: 36208013   Gas used: 589631 (88.66%)   Gas price: 35.03[75](https://github.com/oceanprotocol/ocean.py/actions/runs/3572558250/jobs/6005583795#step:7:76)4352 gwei
```

### Appendix: Simple trace: error 2

```text
-> tests/integration/remote/util.py:104: in do_ocean_tx_and_handle_gotchas
-> assert data_nft.symbol() == symbol
-> /opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/brownie/network/contract.py:1902: in __call__
    return self.call(*args, block_identifier=block_identifier, override=override)

Returns error:
ValueError: No data was returned - the call likely reverted
```

### Appendix: Full trace: error 2

From [this CI log](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656):

```text
________________________ test_ocean_tx__create_data_nft ________________________

tmp_path = PosixPath('/tmp/pytest-of-runner/pytest-0/test_ocean_tx__create_data_nft1')

    @pytest.mark.integration
    def test_ocean_tx__create_data_nft(tmp_path):
        """On Polygon, do a simple Ocean tx: create_data_nft"""
        # setup
        connect_to_network("polygon")
        util.set_aggressive_gas_fees()
    
        config = util.remote_config_polygon(tmp_path)
        ocean = Ocean(config)
    
        accounts.clear()
        (alice_wallet, _) = util.get_wallets()
    
        # Do a simple-as-possible test that uses ocean stack, while accounting for gotchas
>       util.do_ocean_tx_and_handle_gotchas(ocean, alice_wallet)

tests/integration/remote/test_polygon.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/integration/remote/util.py:104: in do_ocean_tx_and_handle_gotchas
    assert data_nft.symbol() == symbol
/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/brownie/network/contract.py:1902: in __call__
    return self.call(*args, block_identifier=block_identifier, override=override)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ContractCall 'symbol()'>, block_identifier = None, override = None
args = ()
tx = {'allow_revert': None, 'data': '0x95d89b41', 'from': None, 'gas': None, ...}

    def call(
        self, *args: Tuple, block_identifier: Union[int, str, bytes] = None, override: Dict = None
    ) -> Any:
        """
        Call the contract method without broadcasting a transaction.
    
        Arguments
        ---------
        *args
            Contract method inputs. You can optionally provide a
            dictionary of transaction properties as the last arg.
        block_identifier : int | str | bytes, optional
            A block number or hash that the call is executed at. If not given, the
            latest block used. Raises `ValueError` if this value is too far in the
            past and you are not using an archival node.
        override : dict, optional
            A mapping from addresses to balance, nonce, code, state, stateDiff
            overrides for the context of the call.
    
        Returns
        -------
            Contract method return value(s).
        """
    
        args, tx = _get_tx(self._owner, args)
        if tx["from"]:
            tx["from"] = str(tx["from"])
        del tx["required_confs"]
        tx.update({"to": self._address, "data": self.encode_input(*args)})
    
        try:
            data = web3.eth.call({k: v for k, v in tx.items() if v}, block_identifier, override)
        except ValueError as e:
            raise VirtualMachineError(e) from None
    
        selector = HexBytes(data)[:4].hex()
    
        if selector == "0x08c379a0":
            revert_str = eth_abi.decode_abi(["string"], HexBytes(data)[4:])[0]
            raise ValueError(f"Call reverted: {revert_str}")
        elif selector == "0x4e487b71":
            error_code = int(HexBytes(data)[4:].hex(), 16)
            if error_code in SOLIDITY_ERROR_CODES:
                revert_str = SOLIDITY_ERROR_CODES[error_code]
            else:
                revert_str = f"Panic (error code: {error_code})"
            raise ValueError(f"Call reverted: {revert_str}")
        if self.abi["outputs"] and not data:
>           raise ValueError("No data was returned - the call likely reverted")
E           ValueError: No data was returned - the call likely reverted

/opt/hostedtoolcache/Python/3.8.14/x64/lib/python3.8/site-packages/brownie/network/contract.py:1708: ValueError
----------------------------- Captured stdout call -----------------------------
alice_wallet.address = '0x8bE6c8126C4383745c25F3b3Ca76D73E3Caf8544'
bob_wallet.address = '0xD[67](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656#step:7:68)64a9F3387B14[69](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656#step:7:70)2CC783BcF5604a23EBb779f'
Call create_data_nft(), and wait for it to complete...

Transaction sent: 0x92808e3ef0dce98a9cf2f299febc283b0172586d3c96cf61b071fbc0e3e9[70](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656#step:7:71)5c
  Max fee: 30.008[71](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656#step:7:72)0186 gwei   Priority fee: 30.0 gwei   Gas limit: 6650[75](https://github.com/oceanprotocol/ocean.py/actions/runs/3572951251/jobs/6006443656#step:7:76)   Nonce: 167
  Waiting for confirmation... -
  Waiting for confirmation... \
  Waiting for confirmation... |
```

### Appendix: Full trace: error 3
```text
(venv) trentmc@trentmc-xps:~/code/ocean.py$ pytest --count=100 -x tests/integration/remote/test_polygon.py::test_ocean_tx__create_data_nft
========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.10.6, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/trentmc/code/ocean.py, configfile: pytest.ini
plugins: env-0.6.2, repeat-0.9.1, eth-brownie-1.19.1, xdist-1.34.0, forked-1.4.0, web3-5.30.0, hypothesis-6.27.3
collected 100 items                                                                                                                                                                                      

tests/integration/remote/test_polygon.py F

================================================================================================ FAILURES ================================================================================================
_________________________________________________________________________________ test_ocean_tx__create_data_nft[1-100] __________________________________________________________________________________

tmp_path = PosixPath('/tmp/pytest-of-trentmc/pytest-20/test_ocean_tx__create_data_nft0')

    @pytest.mark.integration
    def test_ocean_tx__create_data_nft(tmp_path):
        """On Polygon, do a simple Ocean tx: create_data_nft"""
        # setup
        connect_to_network("polygon")
        util.set_aggressive_gas_fees()
    
        config = util.remote_config_polygon(tmp_path)
        ocean = Ocean(config)
    
        accounts.clear()
        (alice_wallet, _) = util.get_wallets()
    
        # Do a simple-as-possible test that uses ocean stack, while accounting for gotchas
>       util.do_ocean_tx_and_handle_gotchas(ocean, alice_wallet)

tests/integration/remote/test_polygon.py:32: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/integration/remote/util.py:102: in do_ocean_tx_and_handle_gotchas
    raise (e)
tests/integration/remote/util.py:96: in do_ocean_tx_and_handle_gotchas
    data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
venv/lib/python3.10/site-packages/enforce_typing/decorator.py:29: in wrapper
    return func(*args, **kwargs)
ocean_lib/ocean/ocean.py:143: in create_data_nft
    receipt = nft_factory.deployERC721Contract(
venv/lib/python3.10/site-packages/brownie/network/contract.py:1861: in __call__
    return self.transact(*args)
venv/lib/python3.10/site-packages/brownie/network/contract.py:1734: in transact
    return tx["from"].transfer(
venv/lib/python3.10/site-packages/brownie/network/account.py:644: in transfer
    receipt, exc = self._make_transaction(
venv/lib/python3.10/site-packages/brownie/network/account.py:752: in _make_transaction
    exc = VirtualMachineError(e)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = VirtualMachineError(ValueError({'code': -32000, 'message': 'nonce too low'})), exc = {'code': -32000, 'message': 'nonce too low'}

    def __init__(self, exc: ValueError) -> None:
        self.txid: str = ""
        self.source: str = ""
        self.revert_type: str = ""
        self.pc: Optional[int] = None
        self.revert_msg: Optional[str] = None
        self.dev_revert_msg: Optional[str] = None
    
        try:
            exc = exc.args[0]
        except Exception:
            pass
    
        if not (isinstance(exc, dict) and "message" in exc):
            raise ValueError(str(exc)) from None
    
        if "data" not in exc:
>           raise ValueError(exc["message"]) from None
E           ValueError: nonce too low

venv/lib/python3.10/site-packages/brownie/exceptions.py:96: ValueError
------------------------------------------------------------------------------------------ Captured stdout call ------------------------------------------------------------------------------------------
alice_wallet.address = '0x8bE6c8126C4383745c25F3b3Ca76D73E3Caf8544'
bob_wallet.address = '0xD6764a9F3387B14692CC783BcF5604a23EBb779f'
Call create_data_nft(), and wait for it to complete...
======================================================================================== short test summary info =========================================================================================
FAILED tests/integration/remote/test_polygon.py::test_ocean_tx__create_data_nft[1-100] - ValueError: nonce too low
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================================================================================== 1 failed in 7.64s ============================================================================================
```

--- a/tests/integration/remote/util.py
+++ b/tests/integration/remote/util.py
@@ -8,14 +8,12 @@ import string
 import time
 import warnings
 
-from brownie.exceptions import ContractNotFound, TransactionError, \
-    VirtualMachineError
+from brownie.exceptions import ContractNotFound, TransactionError, VirtualMachineError
 from brownie.network import accounts, chain, priority_fee
 from enforce_typing import enforce_types
 
 
-ERRORS_TO_CATCH = (ContractNotFound, TransactionError, ValueError,
-                   VirtualMachineError)
+ERRORS_TO_CATCH = (ContractNotFound, TransactionError, ValueError, VirtualMachineError)
 
 
 @enforce_types
@@ -103,7 +101,7 @@ def do_ocean_tx_and_handle_gotchas(ocean, alice_wallet):
     # Alice publish data NFT
     # avoid "replacement transaction underpriced" error: make each tx diff't
     symbol = random_chars()
-    
+
     print("Call create_data_nft(), and wait for it to complete...")
     try:
         data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
@@ -113,20 +111,22 @@ def do_ocean_tx_and_handle_gotchas(ocean, alice_wallet):
             warnings.warn(UserWarning(f"Warning: EVM reported error: {e}"))
             return
         raise (e)
-    
+
     assert data_nft_symbol == symbol
     print("Success")
 
 
 @enforce_types
 def error_is_skippable(error_s: str) -> bool:
-    return ("insufficient funds" in error_s \
-            or "underpriced" in error_s \
-            or "No contract deployed at" in error_s \
-            or "nonce too low" in error_s \
-            or "Internal error" in error_s \
-            or "No data was returned - the call likely reverted" in error_s)
-    
+    return (
+        "insufficient funds" in error_s
+        or "underpriced" in error_s
+        or "No contract deployed at" in error_s
+        or "nonce too low" in error_s
+        or "Internal error" in error_s
+        or "No data was returned - the call likely reverted" in error_s
+    )
+
 
 @enforce_types
 def random_chars() -> str:

--- a/tests/integration/remote/util.py
+++ b/tests/integration/remote/util.py
@@ -8,10 +8,17 @@ import string
 import time
 import warnings
 
-from brownie.exceptions import VirtualMachineError
+from brownie.exceptions import ContractNotFound, TransactionError, \
+    VirtualMachineError
 from brownie.network import accounts, chain, priority_fee
+from enforce_typing import enforce_types
 
 
+ERRORS_TO_CATCH = (ContractNotFound, TransactionError, ValueError,
+                   VirtualMachineError)
+
+
+@enforce_types
 def remote_config_mumbai(tmp_path):
     config = {
         "NETWORK_NAME": "mumbai",
@@ -23,6 +30,7 @@ def remote_config_mumbai(tmp_path):
     return config
 
 
+@enforce_types
 def remote_config_polygon(tmp_path):
     config = {
         "NETWORK_NAME": "polygon",
@@ -34,6 +42,7 @@ def remote_config_polygon(tmp_path):
     return config
 
 
+@enforce_types
 def get_wallets():
     alice_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY1")
     bob_private_key = os.getenv("REMOTE_TEST_PRIVATE_KEY2")
@@ -52,11 +61,13 @@ def get_wallets():
     return (alice_wallet, bob_wallet)
 
 
+@enforce_types
 def set_aggressive_gas_fees():
     # Polygon & Mumbai uses EIP-1559. So, dynamically determine priority fee
     priority_fee(chain.priority_fee)
 
 
+@enforce_types
 def do_nonocean_tx_and_handle_gotchas(ocean, alice_wallet, bob_wallet):
     """Call wallet.transfer(), but handle several gotchas for this test use case:
     - if the test has to repeat, there are nonce errors. Avoid via unique
@@ -65,23 +76,24 @@ def do_nonocean_tx_and_handle_gotchas(ocean, alice_wallet, bob_wallet):
     """
     # Simplest possible tx: Alice send Bob some fake MATIC
     bob_eth_before = accounts.at(bob_wallet.address).balance()
-
     normalized_unixtime = time.time() / 1e9
     amt_send = 1e-8 * (random.random() + normalized_unixtime)
 
     print("Do a send-Ether tx...")
-    try:  # it can get away with some errors, but not others
+    try:
         alice_wallet.transfer(bob_wallet.address, f"{amt_send:.15f} ether")
-    except (ValueError, VirtualMachineError) as e:
-        if "insufficient funds" in str(e) or "underpriced" in str(e):
+        bob_eth_after = accounts.at(bob_wallet.address).balance()
+    except ERRORS_TO_CATCH as e:
+        if error_is_skippable(str(e)):
             warnings.warn(UserWarning(f"Warning: EVM reported error: {e}"))
             return
         raise (e)
 
-    bob_eth_after = accounts.at(bob_wallet.address).balance()
     assert bob_eth_after > bob_eth_before
+    print("Success")
 
 
+@enforce_types
 def do_ocean_tx_and_handle_gotchas(ocean, alice_wallet):
     """Call create_data_nft(), but handle several gotchas for this test use case:
     - if the test has to repeat, there are nonce errors. Avoid via unique
@@ -91,20 +103,32 @@ def do_ocean_tx_and_handle_gotchas(ocean, alice_wallet):
     # Alice publish data NFT
     # avoid "replacement transaction underpriced" error: make each tx diff't
     symbol = random_chars()
-    try:  # it can get away with some errors, but not others
-        print("Call create_data_nft(), and wait for it to complete...")
+    
+    print("Call create_data_nft(), and wait for it to complete...")
+    try:
         data_nft = ocean.create_data_nft(symbol, symbol, alice_wallet)
-
-    except (ValueError, VirtualMachineError) as e:
-        if "insufficient funds" in str(e) or "underpriced" in str(e):
+        data_nft_symbol = data_nft.symbol()
+    except ERRORS_TO_CATCH as e:
+        if error_is_skippable(str(e)):
             warnings.warn(UserWarning(f"Warning: EVM reported error: {e}"))
             return
         raise (e)
-
-    assert data_nft.symbol() == symbol
+    
+    assert data_nft_symbol == symbol
     print("Success")
 
 
+@enforce_types
+def error_is_skippable(error_s: str) -> bool:
+    return ("insufficient funds" in error_s \
+            or "underpriced" in error_s \
+            or "No contract deployed at" in error_s \
+            or "nonce too low" in error_s \
+            or "Internal error" in error_s \
+            or "No data was returned - the call likely reverted" in error_s)
+    
+
+@enforce_types
 def random_chars() -> str:
     cand_chars = string.ascii_uppercase + string.digits
     s = "".join(random.choices(cand_chars, k=8)) + str(time.time())

